### PR TITLE
podman_network_cni: added connect, disconnect and reload commands to be validated.

### PR DESCRIPTION
**Pull request by mgrossu**
- Related ticket: https://progress.opensuse.org/issues/124619
- Verification run: https://openqa.suse.de/tests/11223386 SLE-15-SP5 x86_64, https://openqa.suse.de/tests/11223385 for SLE-15-SP4 x86_64, https://openqa.suse.de/tests/11223387 for SLE-15-SP4 aarch64, https://openqa.suse.de/tests/11223326 for SLE-15-SP1 x86_64, https://openqa.opensuse.org/tests/3326398#step/podman_network_cni/1 for container-host, https://openqa.opensuse.org/tests/3326399#step/podman_network_cni/1 for containers_host_podman

### CODE REVIEW

**Review by danishprakash — APPROVED** [review]
LGTM

File: tests/containers/podman_network.pm [hunk 1]
```diff
@@ -52,10 +52,36 @@ sub run() {
         assert_script_run('podman network inspect newnet1 --format "{{range .Subnets}}Subnet: {{.Subnet}} Gateway: {{.Gateway}}{{end}}"');
     }
 
+    #connect, disconnect & reload
+    unless ($old_podman) {
```
Comment (grisu48) [inline tests/containers/podman_network.pm:56]: `$old_podman` is not really descriptive. Petition to rename it to `$supports_network` or something similar, which is more descriptive.

Why? `old` can mean anything, `supports_network` suggests already what the purpose of the variable is.

**Review by grisu48 — COMMENTED** [review]
Good starting point! I would recommend to add also a functional test of the network from within the container and from outside. We should check, if the container can reach the network and if published ports from the container can be reached for a good integration test.

Overall good job Marius!

**Comment by jlausuch** [comment]
Agree with Felix's comments. Let's try to implement a test close to what a user would do

**Comment by jlausuch** [comment]
Ooops, sorry, I closed it unintentionally...

File: tests/containers/podman_network.pm [hunk 2]
```diff
@@ -52,10 +52,36 @@ sub run() {
...
+        validate_script_output('podman inspect --format="{{.NetworkSettings.Networks}}" container1', sub { m/newnet1/ });
+        validate_script_output('podman inspect --format="{{.NetworkSettings.Networks}}" container2', sub { m/newnet2/ });
+        validate_script_output('podman inspect --format="{{.NetworkSettings.Networks}}" container3', sub { m/newnet2/ });
+
+        record_info('Disconnect', 'Disconnect the container from the network');
+        assert_script_run('podman network disconnect newnet2 container2');
+        validate_script_output('podman inspect --format="{{.NetworkSettings.Networks}}" container2', sub { !m/newnet2/ });
+
+        record_info('Reload', 'Reload the container');
+        validate_script_output('podman network reload container3', sub { m/$container_id/ });
```
[the remark this answers is not in the document]
Comment (danishprakash) [reply tests/containers/podman_network.pm:78]: Quoting from man, this command reloads the container network configuration. For e.g. in rootless podman, which relies on iptables rules to provide network connectivity, the rules get deleted for some reason which would mean the container would lose network connectivity. This command would restore the settings and with it, network connectivity.

[the remark this answers is not in the document]
Comment (danishprakash) [reply tests/containers/podman_network.pm:78]: @mgrossu It'd be nice to delete some iptables rule for the newly created network, test no network connectivity and then reload and then test again to ensure things are working fine. What do you think?

[the remark this answers is not in the document]
Comment (mgrossu) [reply tests/containers/podman_network.pm:78]: @danishprakash Yes, I think it is a good approach; also, in my mind, comes this example from podman docs: https://docs.podman.io/en/latest/markdown/podman-network-reload.1.html  where with firewall-cmd --reload container loses network connectivity

**Comment by pdostal** [comment]
Looks good to me :smile:

File: tests/containers/podman_network.pm (hunk 2 above)
[the remark this answers is not in the document]
Comment (pdostal) [reply tests/containers/podman_network.pm:78]: @mgrossu 
```diff
- record_info('Reload', 'Reload the container');
+ record_info('Reload', 'Reload the container network configuration');
```
please :smile:

**Review by pdostal — APPROVED** [review]

File: tests/containers/podman_network.pm (hunk 2 above)
[the remark this answers is not in the document]
Comment (danishprakash) [reply tests/containers/podman_network.pm:78]: @mgrossu yes, that sounds good. You can implement it as part of another PR while this covers the basic networking testing. Your call!

[the remark this answers is not in the document]
Comment (mgrossu) [reply tests/containers/podman_network.pm]: Ok, i will implement it as another PR.

File: tests/containers/podman_network_cni.pm [hunk 3]
```diff
@@ -52,10 +52,36 @@ sub run() {
...
 
+    #connect, disconnect & reload
+    unless ($old_podman) {
+        record_info('Prepare', 'Prepare three containers');
+        assert_script_run('podman run -id --rm --name container1 -p 1234:1234 registry.opensuse.org/opensuse/tumbleweed');
+        assert_script_run('podman run -id --rm --name container2 -p 1235:1235 registry.opensuse.org/opensuse/tumbleweed');
+        my $container_id = assert_script_run('podman run -id --rm --name container3 -p 8080:80 registry.opensuse.org/opensuse/nginx');
+
+
+        record_info('Connect', 'Connect the container to the network');
```
Comment (danishprakash) [inline tests/containers/podman_network_cni.pm:63]:
```suggestion
        record_info('Connect', 'Connect the containers to the networks');
```

**Review by danishprakash — APPROVED** [review]
LGTM

File: tests/containers/podman_network_cni.pm (hunk 3 above)
Comment (mgrossu) [reply tests/containers/podman_network_cni.pm]: Done.

**Review by ricardobranco777 — APPROVED** [review]
I would have used a much smaller image like `registry.opensuse.org/bci/bci-busybox:latest` instead of TW.

LGTM.

File: tests/containers/podman_network_cni.pm [hunk 4]
```diff
@@ -52,10 +52,36 @@ sub run() {
         assert_script_run('podman network inspect newnet1 --format "{{range .Subnets}}Subnet: {{.Subnet}} Gateway: {{.Gateway}}{{end}}"');
     }
 
+    #connect, disconnect & reload
+    unless ($old_podman) {
+        record_info('Prepare', 'Prepare three containers');
+        assert_script_run('podman run -id --rm --name container1 -p 1234:1234 registry.opensuse.org/opensuse/tumbleweed');
```
Comment (jlausuch) [inline tests/containers/podman_network_cni.pm:58]: from my experience with pulling containers, I would first run the command `podman pull registry.opensuse.org/opensuse/tumbleweed` using `script_retry` before using the image. 
Example: https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/tests/containers/bci_repo.pm#L82

File: tests/containers/podman_network_cni.pm [hunk 5]
```diff
@@ -52,10 +52,36 @@ sub run() {
...
+        assert_script_run('podman run -id --rm --name container2 -p 1235:1235 registry.opensuse.org/opensuse/tumbleweed');
+        my $container_id = assert_script_run('podman run -id --rm --name container3 -p 8080:80 registry.opensuse.org/opensuse/nginx');
```
Comment (mloviska) [inline tests/containers/podman_network_cni.pm:60]: Isn't `$container_id` empty?

Comment (grisu48) [reply tests/containers/podman_network_cni.pm:60]: Yes, likely it should be `script_output`. See https://open.qa/api/testapi/#_assert_script_run and https://open.qa/api/testapi/#_script_output

Comment (mgrossu) [reply tests/containers/podman_network_cni.pm]: Fixed.

File: tests/containers/podman_network_cni.pm (hunk 4 above)
Comment (mgrossu) [reply tests/containers/podman_network_cni.pm:63]: @jlausuch Your suggestions was implemented.

**Comment by jlausuch** [comment]
Please add also VRs from openSUSE, containers_host_podman and microos container-host flavor.

**Review by jlausuch — APPROVED** [review]
LGTM.
For the next PR, please change `$old_podman` by `$supports_network` as Felix suggests.

**Comment by rfan1** [comment]
Hello @mgrossu seems leap tests are failed https://openqa.opensuse.org/tests/3329997#step/podman_network_cni/129

**Pull request merged**